pkg/operator/upgradebackupcontroller/upgradebackupcontroller.go

```diff
@@ -0,0 +1,258 @@
+	delete := func(pod *corev1.Pod) error {
+		err := kubeClient.CoreV1().Pods(operatorclient.TargetNamespace).Delete(ctx, pod.Name, metav1.DeleteOptions{Preconditions: metav1.NewUIDPreconditions(string(pod.UID))})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	backupPods, err := podLister.List(labels.Set{"app": "cluster-backup"}.AsSelector())
+	if err != nil {
+		return "", err
+	}
+
+	var errs []error
+	for _, pod := range backupPods {
+		if pod.Labels[clusterBackupVersionLabel] != clusterBackupVersion {
+			// Delete pods with invalid version
+			if err := delete(pod); err != nil {
+				errs = append(errs, err)
+			}
+			continue
+		}
+
+		switch pod.Status.Phase {
+		case corev1.PodSucceeded:
+			return operatorv1.ConditionTrue, nil
+		case corev1.PodFailed:
+			if err := delete(pod); err != nil {
+				klog.Errorf("failed to delete pod %s/%s: %v", pod.Namespace, pod.Name, err)
+			}
+			return operatorv1.ConditionFalse, nil
+		default:
+			return operatorv1.ConditionUnknown, nil
+		}
+	}
+
+	node, err := getNodeWithAdequateAllocatableStorage(ctx, podLister, nodeLister)
+	if err != nil {
+		return "", err
+	}
+
+	_, _, err = manageRecoveryPod(ctx, node.Name, targetImagePullSpec, clusterBackupVersion, kubeClient.CoreV1(), recorder)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("%q: %v", "pod/cluster-backup", err))
+	}
+
+	if len(errs) > 0 {
+		return "", utilerrors.NewAggregate(errs)
+	}
+
+	return operatorv1.ConditionUnknown, nil
+}
+
+func manageRecoveryPod(ctx context.Context, nodeName, targetImagePullSpec, clusterBackupVersion string, client coreclientv1.PodsGetter, recorder events.Recorder) (*corev1.Pod, bool, error) {
+	pod := resourceread.ReadPodV1OrDie(etcd_assets.MustAsset("etcd/cluster-backup-pod.yaml"))
+	pod.Name = "cluster-backup-" + nodeName
+	pod.Namespace = operatorclient.TargetNamespace
+	pod.Spec.NodeName = nodeName
+	pod.Labels = map[string]string{
+		"app":                     "cluster-backup",
+		clusterBackupVersionLabel: clusterBackupVersion,
+	}
+	pod.Spec.Containers[0].Image = targetImagePullSpec
+
+	return resourceapply.ApplyPod(ctx, client, recorder, pod)
+}
+
+// isRequireRecentBackup checks the conditions of ClusterVersion to verify if a backup is required.
+func isRequireRecentBackup(config *configv1.ClusterVersion) bool {
+	for _, condition := range config.Status.Conditions {
+		// Check if Failing is true and Message field containers the string RecentBackup.
+		if condition.Type == "Failing" && condition.Status == configv1.ConditionTrue {
+			return strings.Contains(condition.Message, "RecentBackup") == true
+		}
+	}
+	return false
+}
+
+// getNodeWithAdequateAllocatableStorage returns a control-plane node that meet the minimum storage requirements for the backup state.
+func getNodeWithAdequateAllocatableStorage(_ context.Context, podLister corev1listers.PodLister, nodeLister corev1listers.NodeLister) (*corev1.Node, error) {
```

Comment:
I guess the plan is to use ctx in the future? I don't see it being needed right now by any of the existing functions?

Reply:
yeah wasn't sure here if I was going to ask etcd about backend status.